azkaban-common/src/main/java/azkaban/executor/container/watch/KubernetesWatch.java

```diff
@@ -0,0 +1,247 @@
+    // no timeout for request completion
+    OkHttpClient httpClient =
+        this.apiClient.getHttpClient().newBuilder().readTimeout(0, TimeUnit.SECONDS).build();
+    this.apiClient.setHttpClient(httpClient);
+    this.coreV1Api = new CoreV1Api(this.apiClient);
+
+    watchRunner = new Thread(this::initializeAndStartPodWatch);
+  }
+
+  public int getPodWatchInitCount() {
+    return podWatchInitCount;
+  }
+
+  @VisibleForTesting
+  protected void setPodWatch(Watch<V1Pod> podWatch) {
+    requireNonNull(podWatch, "pod watch must not be null");
+    this.podWatch = podWatch;
+  }
+
+  /**
+   * Create the Pod watch and set it up for creating parsed representations of the JSON
+   * responses received from the Kubernetes API server. Responses will be converted to type
+   * {@code Watch.Response<V1Pod>}.
+   * Creating the watch submits the request the API server but does not block beyond that.
+   *
+   * @throws ApiException
+   */
+  protected void initializePodWatch() throws ApiException {
+    try {
+      this.podWatch = Watch.createWatch(this.apiClient,
+          coreV1Api.listNamespacedPodCall(podWatchParams.getNamespace(),
+              "true",
+              false,
+              null,
+              null,
+              podWatchParams.getLabelSelector(),
+              null,
+              null,
+              null,
+              true,
+              null),
+          new TypeToken<Response<V1Pod>>() {}.getType());
+    } catch (ApiException ae) {
+      logger.error("ApiException while creating pod watch.", ae);
+      throw ae;
+    }
+  }
+
+  /**
+   * This starts the continuous event processing loop for fetching the pod watch events.
+   * Processing of the events is callback driven and the registered {@link RawPodWatchEventListener}
+   * provides the processing implementation.
+   *
+   * @throws IOException
+   */
+  protected void startPodWatch() throws IOException {
+    requireNonNull(podWatch, "watch must be initialized");
+    for (Watch.Response<V1Pod> item : podWatch) {
+      if (isShutdownRequested.get()) {
+        logger.info("Exiting pod watch event loop as shutdown was requested");
+        return;
+      }
+      this.podWatchEventListener.onEvent(item);
+    }
+  }
+
+  private void closePodWatchQuietly() {
+    if (podWatch == null) {
+      logger.debug("Pod watch is null");
+      return;
+    }
+    try {
+      podWatch.close();
+    } catch (IOException e) {
+      logger.error("IOException while closing pod watch.", e);
+    }
+  }
+
+  /**
+   * Initialize and start the Pod watch procedure. The processing is expected to continue until a
+   * shutdown request is received. The method will handle any exceptions received during both (1)
+   * the execution of kubernetes API calls (2) processing of events received by the watch.
+   * This also includes re-initialization of the watch using Kuberentes watch API, as needed.
+   */
+  public void initializeAndStartPodWatch() {
```

Comment:
Shall it be thread safe?

Reply:
Fixed. This method was made private and synchronization with additional checks added in `launchPodWatch`